src/libraries/System.Private.CoreLib/src/System/Numerics/TotalOrderIeee754Comparer.cs

```diff
@@ -0,0 +1,206 @@
+                // here we define it as the order of exponent bits to simplify comparison
+
+                // Negative values are represented in sign-magnitude, instead of two's complement like integers
+                // Just negating the comparison result when both numbers are negative is enough
+
+                return (TInteger.IsNegative(x) && TInteger.IsNegative(y)) ? y.CompareTo(x) : x.CompareTo(y);
+            }
+
+            static int CompareGeneric(T? x, T? y)
+            {
+                // IComparer contract is null < value
+
+                if (x is null)
+                {
+                    return (y is null) ? 0 : -1;
+                }
+                else if (y is null)
+                {
+                    return 1;
+                }
+
+                // If < or > returns true, the result satisfies definition of totalOrder too
+
+                if (x < y)
+                {
+                    return -1;
+                }
+                else if (x > y)
+                {
+                    return 1;
+                }
+                else if (x == y)
+                {
+                    if (T.IsZero(x)) // only zeros are equal to zeros
+                    {
+                        // IEEE 754 numbers are either positive or negative. Skip check for the opposite.
+
+                        if (T.IsNegative(x))
+                        {
+                            return T.IsNegative(y) ? 0 : -1;
+                        }
+                        else
+                        {
+                            return T.IsPositive(y) ? 0 : 1;
+                        }
+                    }
+                    else
+                    {
+                        // Equivalant values are compared by their exponent parts,
+                        // and the value with smaller exponent is considered closer to zero.
+
+                        // This only applies to IEEE 754 decimals. Consider to add support if decimals are added into .NET.
+                        return 0;
+                    }
+                }
+                else
+                {
+                    // One or two of the values are NaN
+                    // totalOrder defines that -qNaN < -sNaN < x < +sNaN < + qNaN
+
+                    static int CompareSignificant(T x, T y)
+                    {
+                        // IEEE 754 totalOrder only defines the order of NaN type bit (the first bit of significant)
+                        // To match the integer semantic comparison above, here we compare all the significant bits
+                        // Revisit this if decimals are added
+
+                        Span<byte> significantX = stackalloc byte[x!.GetSignificandByteCount()];
+                        Span<byte> significantY = stackalloc byte[y!.GetSignificandByteCount()];
+                        x.WriteSignificandBigEndian(significantX);
+                        y.WriteSignificandBigEndian(significantY);
```

Comment:
This is a potential `StackOverflow` for custom floating-point types (imagine a `BigFloat` that could have more than 1024 bytes in the significand).

This could opt to `GetSignificandBitLength()` and only get the actual bytes if the lengths are equal

Reply:
Notably I imagine most `nans` will have this be `0` and so we can just short-circuit the majority of cases